File: fx/src/io/github/gravetii/controller/GameTimerTask.java
package io.github.gravetii.controller;

import io.github.gravetii.game.Game;
import javafx.concurrent.Task;
import javafx.scene.layout.BorderPane;

public class GameTimerTask extends Task<Void> {

  private BorderPane root;
  private double time;

  public GameTimerTask(BorderPane root, double time) {
    this.root = root;
    this.time = time;
  }

  @Override
  protected Void call() {
    if (this.time == 0) {
      this.withoutLimit();
    } else {
      this.withLimit();
    }

    return null;
  }

  private void withLimit() {
    double c = this.time;
    try {
      while (c > 0) {
        Thread.sleep(50);
        this.updateProgress(c, this.time);
        c -= 0.05;
      }

      this.updateProgress(Double.MIN_VALUE, this.time);
    } catch (InterruptedException e) {
      // do nothing
    }

    this.root.getChildren().forEach(node -> node.fireEvent(new Game.EndEvent()));
  }

  private void withoutLimit() {
    this.updateProgress(1, 1);
  }
}

File: src/main/kotlin/io/github/gravetii/controller/GameTimerTask.kt
package io.github.gravetii.controller

import io.github.gravetii.game.Game
import javafx.concurrent.Task
import javafx.scene.layout.BorderPane
import java.util.logging.Logger

class GameTimerTask(private val root: BorderPane, private val time: Double) : Task<Unit>() {

    private val logger = Logger.getLogger(javaClass.canonicalName)

    override fun call() {
        var c: Double = time
        try {
            while (c > 0) {
                Thread.sleep(50)
                updateProgress(c, time)
                c -= 0.05
            }

            updateProgress(Double.MIN_VALUE, time)
            root.fireEvent(Game.Companion.EndEvent())
        } catch (e: InterruptedException) {
            logger.fine("Game stopped in between")
        }
    }


}